File: solutions/clojure/matrix/1/src/matrix.clj
(ns matrix
  (:require [clojure.string :as s]))

(defn ->matrix [s]
  (->> s
       (s/split-lines)
       (mapv (comp read-string (partial format "[%s]")))))

(defn get-row
  "Returns the i-th row of the matrix s"
  [s i]
  (-> s
     ->matrix
     (get (dec i))))


(defn get-column
  "Returns the i-th column of the matrix s"
  [s i]
  (->> s
       ->matrix
       (apply mapv vector)
       (#(get % (dec i)))))

File: solutions/scala/matrix/1/src/main/scala/Matrix.scala
case class Matrix(input: String) {
  private val rows: Vector[Vector[Int]] =
    input.linesIterator
      .map(_.split(" ").map(_.toInt).toVector)
      .toVector

  def row(index: Int): Vector[Int] =
    rows(index)

  def column(index: Int): Vector[Int] =
    rows.map(_(index))
}
